Seqr dataset cleanup scripts

EddieLF opened the pull request:
**Tacked onto this PR are some updates to the pre-commit hooks and also some very minor linting to some existing files. Also added `-m` flag to the existing `cleanup_seqr_load_files.py` script in the `gsutil rm` section.**

---

New script which searches through the `mt/` and `exome/mt/` paths within the `main` or `test` buckets for the input datasets. 
Any `.mt` folders that are not the latest created matrix table are deleted.

We should delete these files because we do not need them, they are essentially an intermediate output between the AnnotateCohort and MtToEs stages.

Here are some test runs:
https://batch.hail.populationgenomics.org.au/batches/429645/jobs/1
https://batch.hail.populationgenomics.org.au/batches/429648/jobs/1

Difficult to find a `test` project with .mt folders for both genome and exome, so I've attached some logs from a dry-run over a main bucket and uploaded them to google drive [here](https://drive.google.com/file/d/17cO9ySiyUHgFdpcQSRFYn3yBH54Jwb71/view?usp=sharing).

EddieLF committed:
Script to cleanup .mt files from AnnotateDataset

EddieLF committed:
Linting updates for existing scripts

EddieLF committed:
Remove unused variable from cleanup matrixtable script

EddieLF committed:
Catch errors if no mt folders found

EddieLF committed:
Add type annotation for mt_timestamp dict

EddieLF committed:
Remove -f flag from gsutil rm in favour of -m flag

cassimons reviewed (commented):
Thanks Ed, looks great, but I do not have time to walk through the logic as much as like to for these deletion scripts. @MattWellie are you up to do the review on this?

MattWellie commented:
I'll have a look!

@EddieLF can you also add a bump to these github actions, both have an `@v4` version available

EddieLF committed:
Bump github actions versions

MattWellie commented on data_management_scripts/cleanup_seqr_loader_files.py:
```diff
@@ -75,8 +75,8 @@ def delete_seqr_load_files(seqr_loads_to_delete: list[str]):
             'rm',
-            '-f',
             '-r',
             *seqr_loads_to_delete,
```
if the image supports it this could be a `gcloud storage` command (which is parallel-enabled by default). It _should_ have gcloud, the lib has been available for over a year and makes many GCP operations much faster.

`gcloud storage rm --recursive *seqr_loads_to_delete`

MattWellie commented on data_management_scripts/cleanup_dataset_matrixtables.py:
```diff
@@ -0,0 +1,150 @@
+):
+    """Finds the .mt directories and their create times in the seqr-main bucket"""
+    mt_timestamps: dict[str, datetime] | None = {}
+    try:
```
Instead of a subprocess call and string parsing, this can be done more cleanly (I think) with cloudpathlib

```
MT_SUFFIXES = [GENOME_PREFIX, EXOME_PREFIX]
bucket = get_config()['storage'][dataset]['default']
for suffix in MT_SUFFIXES:
    mt_folder = to_path(join(bucket, suffix))
    if not mt_folder.exists():
        continue  # do something else
    mt_files = list(mt_folder.glob('*.mt')) # cast generator to list
    # list(mt_folder.iterdir()) would do the same thing, and is also near-instant
```

MattWellie replied:
the timestamps _should_ be recoverable as UNIX timestamps from the GSPath objects - running a little experiment I can see the `st_mtime` (modified time?) but all other attributes are None... hmm

```
>> all_mts = [each_path / '_SUCCESS' for each_path in mt_files]
>> t = all_mts[0]
>> t
'gs://cpg-acute-care-main/exome/mt/3c9acca7967c89b19386d897d14c1add532ebb_1870-acute-care.mt/_SUCCESS'
>> t.stat()
os.stat_result(st_mode=None, st_ino=None, st_dev='gs://', st_nlink=None, st_uid=None, st_gid=None, st_size=0, st_atime=None, st_mtime=1682459118.557, st_ctime=None)
>> t.stat().st_mtime
1682459118.557
```

MattWellie commented on data_management_scripts/cleanup_dataset_matrixtables.py:
```diff
@@ -0,0 +1,150 @@
+EXOME_PREFIX = 'exome/mt/'
+
+
+def get_dataset_mt_timestamps(
```
I very much thought this should be coming from Metamist. I was today years old when I realised that we've never registered these as Analysis entries 😞

MattWellie commented on data_management_scripts/cleanup_dataset_matrixtables.py:
```diff
@@ -0,0 +1,150 @@
+    return dataset_mts_to_delete
+
+
+def delete_dataset_mts(dataset: str, dataset_mts_to_delete: list[str]):
```
always makes me feel squiffy submitting a job that intends to delete a subset of the main project data, and determines that subset at runtime...

I would feel more comfortable splitting this into a script to find paths to delete (which could be run locally) -> generate a bash script of gcloud/gsutil delete commands -> double check -> submit that job. My appetite for risk isn't what it used to be

MattWellie commented on data_management_scripts/cleanup_dataset_matrixtables.py:
```diff
@@ -0,0 +1,150 @@
+    mt_timestamps_dict: dict[str, datetime],
+):
+    """Takes the list of .mt directories and keeps only the most recently created one"""
+    dataset_mts = sorted(mt_timestamps_dict.items(), key=lambda x: x[1])
```
Sorting by timestamp is fine, but we should also be happy ripping this from the names. The MT naming convention is `BUCKET/mt/RANDOM-HASH_NUMBER-OF-SAMPLES-PROJECT.mt` - where presumably the latest will always have the highest number of included samples

MattWellie reviewed (changes requested):
I think this will act exactly as intended, but I don't feel comfortable with it. I would much rather split into

1. script to find deletion candidates
2. bash script to remove those candidates (Submitted via PR for auditability or what we are deleting - it's a real shame these results weren't registered in metamit)

Deletion of a list of objects determined at runtime scares me

cassimons commented:
> I think this will act exactly as intended, but I don't feel comfortable with it. I would much rather split into
> 
> script to find deletion candidates
> bash script to remove those candidates (Submitted via PR for auditability or what we are deleting - it's a real shame these results weren't registered in metamit)
> Deletion of a list of objects determined at runtime scares me

I understand and in most cases agree with this sentiment, particularly for one-off or rare data deletion operations or data types that are expensive or hard to regenerate. However, we need to balance the risk against the cost we incur by keeping multiple copies of these files in storage. 

These MTs are large, we currently generate new ones at least once a month and keep all copies until Ed does a cleanout. We also keep the full cohort MT and an effective copy of these in Elastic. Given dataset MTs are relatively cheap to re-generate from the court MT we should probably consider if these should be written to tmp. Unless we go to that model then we need to enable a largely hands-off process (ideally run by automation) to aggressively prune historic outputs without requiring a multistep manual process by Ed.

That may mean we need to put extra work into the script to make sure we are happy it will be safe, but it will be with the effort in both storage cost and Ed's mental load.

EddieLF committed:
Remove gsutil usage and improve sequencing type iteration

EddieLF committed:
Remove subprocess usage, use CloudPath.rmtree() instead

EddieLF committed:
Remove a file that was renamed and re-added

EddieLF commented:
@MattWellie I've come back to this and made some updates! To summarize:

**General**
- Use `CloudPath.rmtree()` to recursively remove `.mt/` folders and their contents. This avoids using subprocesses.
- Use `--dry-run` flag to print the items to delete without deleting them.

**Scripts**

**`cleanup_seqr_loader_files.py`**
- Searches the `seqr_loader/` and `exome/seqr_loader/` in the `gs://cpg-seqr-main/` bucket for .mt/ folders created by `AnnotateCohort`. 
- Keeps the most recently created folder & any created in the last 30 days. Deletes the remaining folders once the dry run flag is off.
- Once `AnnotateCohort` entries are their own thing in Metamist, this script should be archived and replaced by one that uses Metamist.

**`search_and_delete_bucket_matrixtables.py`**
- A one time use script to find and delete any .mt/ folders in each datasets exome and genome folders created by `AnnotateDataset` and `AnnotateDatasetSV`.
- Keeps only the most recently created folder (agnostic of analysis type, sv or custom), deletes the rest in the absence of the dry run flag.
- Once this has been run, all newly created dataset .mt folders will be logged in metamist, so this script should be archived and replaced by the next one below.

**`cleanup_matrixtable_analysis_outputs.py`**
- Queries metamist for dataset analysis with `type == 'custom'` and `meta.stage == 'AnnotateDataset'`, for both genome and exome.
-  Queries metamist for dataset analysis with `type == 'sv'` and `meta.stage == 'AnnotateDatasetSv'`, for both genome and exome.
- For each of the above combinations, keeps the most recently created .mt folder and deletes the rest when the dry run flag is absent.


I will follow this up with some test outputs from local dry runs. Cheers & have a great weekend.
